### PR TITLE
mds: optimize revoking stale caps

### PR DESCRIPTION
**Pull request by ukernel**
http://tracker.ceph.com/issues/38043

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by batrick** [comment]
@ukernel let me know when this is ready for review.

**Comment by ukernel** [comment]
It's ready

File: src/client/Client.cc [hunk 1]
```diff
@@ -3543,9 +3549,8 @@ void Client::check_caps(Inode *in, unsigned flags)
     if (!revoking && unmounting && (cap_used == 0))
       goto ack;
 
-    if (wanted == cap.wanted &&         // mds knows what we want.
```
Comment (batrick) [inline src/client/Client.cc]: Question regarding this change (7922a292e9d1c69cc2298b0126ae3a16b65c2763): does this mean the MDS will always wait for the revoke of issued caps even if the wanted caps do not include the caps? Is `wanted` really just for communicating the desired caps by the client (i.e. "MDS please give me all of these caps if possible")?

**Comment by batrick** [comment]
I don't think we can do 2bfe1573b94cafad8ed16878c8a1a848fdb8dc1c. Since the unmitigated problem is that a client can cause the MDS to chew through millions of caps over 15+ seconds, resetting the heartbeat is just admitting the MDS is broken. We either need to (a) prevent the MDS from ever getting in this situation by limiting the number of outstanding caps somehow, or (b) write a resumable continuation that revokes the caps incrementally.

**Comment by batrick** [comment]
Adding onto my previous comment (https://github.com/ceph/ceph/pull/25254#issuecomment-449154783), it appears that commit isn't even necessary since the other commits appears to resolve the problem.

**Comment by ukernel** [comment]
"reset heartbeat" commit removed

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
/ceph/teuthology-archive/pdonnell-2019-01-12_05:29:15-fs-wip-pdonnell-testing-20190112.024010-distro-basic-smithi/3452126/teuthology.log

confirmed in a vstart cluster this PR causes it

**Comment by ukernel** [comment]
Fixed
````
diff --git a/src/mds/Capability.cc b/src/mds/Capability.cc
index 080cd76037..eadbd5cdff 100644
--- a/src/mds/Capability.cc
+++ b/src/mds/Capability.cc
@@ -153,8 +153,10 @@ Capability::Capability(CInode *i, Session *s, uint64_t id) :
   _pending(0), _issued(0), last_sent(0), last_issue(0), mseq(0),
   suppress(0), state(0)
 {
-  session->touch_cap_bottom(this);
-  cap_gen = session->get_cap_gen();
+  if (session) {
+    session->touch_cap_bottom(this);
+    cap_gen = session->get_cap_gen();
+  }
 }
 
 client_t Capability::get_client() const
````

**Comment by ukernel** [comment]
add 'Fixes: http://tracker.ceph.com/issues/38043" to the last commit. No other change

**Pull request merged**